flink-table-store-core/src/main/java/org/apache/flink/table/store/table/source/SnapshotEnumerator.java

```diff
@@ -99,11 +103,48 @@ public static class EnumeratorResult {
 
         public final long snapshotId;
 
-        public final DataTableScan.DataFilePlan plan;
+        public final DataFilePlan plan;
 
-        private EnumeratorResult(long snapshotId, DataTableScan.DataFilePlan plan) {
+        private EnumeratorResult(long snapshotId, DataFilePlan plan) {
             this.snapshotId = snapshotId;
             this.plan = plan;
         }
     }
+
+    /** Startup snapshot enumerator, this is the first plan for continuous reading. */
+    public static DataFilePlan startup(DataTableScan scan) {
+        CoreOptions options = scan.options();
+        SnapshotManager snapshotManager = scan.snapshotManager();
+        CoreOptions.LogStartupMode startupMode = options.logStartupMode();
+        switch (startupMode) {
+            case FULL:
+                DataFilePlan plan;
+                if (options.changelogProducer() == FULL_COMPACTION) {
+                    // Read the results of the last full compaction.
+                    // Only full compaction results will appear on the max level.
+                    plan = scan.withLevel(options.numLevels() - 1).plan();
+                } else {
+                    plan = scan.plan();
+                }
+                return plan;
+            case LATEST:
+                return new DataFilePlan(
+                        snapshotManager.latestSnapshotId(), Collections.emptyList());
+            case FROM_TIMESTAMP:
+                Long timestampMills = options.logScanTimestampMills();
+                if (timestampMills == null) {
```

Comment:
Could the `LOG_SCAN_TIMESTAMP_MILLS ` validation move before creating source?

Reply:
I will introduce `validateOptions` in `FileStoreTableFactory`.